Expose `_validate_property()` for scripting

KoBeWi opened the pull request:
`_validate_property()` has (at least) 2 uses in scripting:
- convenient customization of exported property list (you can change usage without using `_get_property_list()`+`_get()`+`_set()`)
- customizing built-in properties (hiding them, disabling etc. Right now it's not possible at all)

This PR exposes it to users. I did it only for GDScript, as I'm not sure how to do it in other languages.
I also added a call to `_validate_property()` for exported properties. This is now possible:
```GDScript
@export var test = 1

func _validate_property(property):
    if property.name == "test":
        property.usage |= PROPERTY_USAGE_READONLY
```

KoBeWi committed:
Expose _validate_property() for scripting

KoBeWi commented:
> the problem with passing property info to extensions remains

Pretty sure it can be done the same way as in `_get_property_list()`. My problem was with how these functions are defined in extensions. It's a spaghetti of method pointer typedefs. I didn't even look at C#, because I could compile without it 🙃

YuriSizov commented:
> Pretty sure it can be done the same way as in `_get_property_list()`.

For me the problem was that with validate it's a two way communication, whereas property list only returns a list of properties. For validate we need to construct `GDExtensionPropertyInfo` from `PropertyInfo`, pass it to extensions, and then somehow handle changes (is it a reference, is it a value?) Which was way beyond my comfort zone to implement. I'm sure it's doable :)

KoBeWi commented:
I changed the calling order. Now the script `_validate_property()` is called after the virtual one, so users can validate already validated properties.

I also added a C# method stub.

akien-mga commented on modules/mono/csharp_script.cpp:
```diff
@@ -1696,6 +1696,10 @@ bool CSharpInstance::property_can_revert(const StringName &p_name) const {
 	return (bool)ret;
 }
 
+void CSharpInstance::validate_property(PropertyInfo &p_property) const {
+	// TODO
+}
```
Maybe @godotengine/dotnet can help with the implementation.

raulsntos replied:
Here's a diff with the implementation (also adds a C# example to the `_validate_property` method documentation):

```diff
diff --git a/doc/classes/Object.xml b/doc/classes/Object.xml
index 90ff9259bc..515a0c98dd 100644
--- a/doc/classes/Object.xml
+++ b/doc/classes/Object.xml
@@ -267,7 +267,8 @@
 			<param index="0" name="property" type="Dictionary" />
 			<description>
 				Override this method to customize existing properties. Every property info goes through this method. The dictionary contents is the same as in [method _get_property_list].
-				[codeblock]
+				[codeblocks]
+				[gdscript]
 				@tool
 				extends Node
 
@@ -280,7 +281,38 @@
 				func _validate_property(property: Dictionary):
 				    if property.name == "number" and not is_number_editable:
 				        property.usage |= PROPERTY_USAGE_READ_ONLY
-				[/codeblock]
+				[/gdscript]
+				[csharp]
+				[Tool]
+				public partial class MyNode : Node
+				{
+				    private bool _isNumberEditable;
+
+				    [Export]
+				    public bool IsNumberEditable
+				    {
+				        get =&gt; _isNumberEditable;
+				        set
+				        {
+				            _isNumberEditable = value;
+				            NotifyPropertyListChanged();
+				        }
+				    }
+
+				    [Export]
+				    public int Number { get; set; }
+
+				    public override void _ValidateProperty(Godot.Collections.Dictionary property)
+				    {
+				        if (property["name"].AsStringName() == PropertyName.Number &amp;&amp; IsNumberEditable)
+				        {
+				            var usage = property["usage"].As&gt;PropertyUsageFlags&lt;() | PropertyUsageFlags.ReadOnly;
+				            property["usage"] = (int)usage;
+				        }
+				    }
+				}
+				[/csharp]
+				[/codeblocks]
 			</description>
 		</method>
 		<method name="add_user_signal">
diff --git a/modules/mono/csharp_script.cpp b/modules/mono/csharp_script.cpp
index 02c5605022..ebc526b864 100644
--- a/modules/mono/csharp_script.cpp
+++ b/modules/mono/csharp_script.cpp
@@ -1658,7 +1658,8 @@ void CSharpInstance::get_property_list(List<PropertyInfo> *p_properties) const {
 		}
 	}
 
-	for (const PropertyInfo &prop : props) {
+	for (PropertyInfo &prop : props) {
+		validate_property(prop);
 		p_properties->push_back(prop);
 	}
 }
@@ -1697,7 +1698,21 @@ bool CSharpInstance::property_can_revert(const StringName &p_name) const {
 }
 
 void CSharpInstance::validate_property(PropertyInfo &p_property) const {
-	// TODO
+	ERR_FAIL_COND(!script.is_valid());
+
+	Variant property_arg = (Dictionary)p_property;
+	const Variant *args[1] = { &property_arg };
+
+	Variant ret;
+	Callable::CallError call_error;
+	GDMonoCache::managed_callbacks.CSharpInstanceBridge_Call(
+			gchandle.get_intptr(), &SNAME("_validate_property"), args, 1, &call_error, &ret);
+
+	if (call_error.error != Callable::CallError::CALL_OK) {
+		return;
+	}
+
+	p_property = PropertyInfo::from_dict(property_arg);
 }
 
 bool CSharpInstance::property_get_revert(const StringName &p_name, Variant &r_ret) const {

```

dsnopek commented on core/object/script_language_extension.h:
```diff
@@ -675,6 +675,9 @@ class ScriptInstanceExtension : public ScriptInstance {
 		}
 		return Variant::NIL;
 	}
+	virtual void validate_property(PropertyInfo &p_property) const override {
+		// TODO
+	}
```
This'll require adding a new function pointer to `GDExtensionScriptInstanceInfo`, which (per the pattern being established in PR https://github.com/godotengine/godot/pull/78634) actually means making a new struct with the new property and a new version of the `script_instance_create` function in `gdextension_interface.h`. We've been holding off on these sort of changes until PR https://github.com/godotengine/godot/pull/78634 is merged, wanting to get at least the first one in before doubling down on this pattern.

In any case, I'd be happy to either write the code to do this (not sure the best way to provide it with this unmerged, though?) or to provide guidance on how to do it.

akien-mga replied:
Makes sense. I guess let's wait for #78634, we could merge this PR with the TODO in the meantime, and make note / open an issue about implementing the GDExtension counterpart before the 4.2 release.

akien-mga reviewed (approved):
Looks good to me.

akien-mga commented:
Thanks!
